### ISSUE TITLE
Consider adding a 3rd / new "status" report type for pledge

### ISSUE THREAD

**Issue by EskoDijk** [anima-wg/anima-brski-prm#35]
In Fig 3, the last thing the pledge does is report its enrollment status which basically checks the LDevID cert is ok.

But even when that's successful, it may be that the pledge is unable to join the domain network with that LDevID cert, later on (when a local domain network becomes available e.g. in 1 month time).  Then the commissioner might have no clue why this new device didn't join in the end.

For troubleshooting purposes, it may be relevant to add a new / 3rd status reporting type that can be used by the registrar-agent to visit the device again and ask for status.

One generic way to report 'errors' or status in general might be [https://datatracker.ietf.org/doc/html/draft-ietf-core-problem-details-06](https://datatracker.ietf.org/doc/html/draft-ietf-core-problem-details-06)

But this could also be vendor-specific and then we could say it's nice to have the additional troubleshooting endpoint on the pledge, but it's out of our scope.

**Comment by stfries** [comment]
Would this lead automatically to a new endpoint at the pledge to allow status requests from a registrar-agent? The currently available status reports are always answers to objects provided (certificate, voucher)

**Comment by EskoDijk** [comment]
Yes, that would be a new endpoint/resource at the pledge. Roughly there would be status info returned indicating one of the phases:
* No LDevID credentials yet
* Have LDevID credentials, but didn't get a chance yet to connect to a network with it   (could include some indicated period of time that the device tried)
* Tried to connect to network, but failed for reason X (more detail in an object reporting format) 
* Success - now connected to network

A small dilemma here is the security - is it wise to leave the device "open" to a visiting tool, once the bootstrap is complete? Maybe only keep the device "open" to the tool if unable to connect to the network?

**Comment by stfries** [comment]
While still thinking about it, one would expect that the pledge only provides return codes indicating the success/failure of an operation. To query the status of the pledge for an arbitrary caller would most likely something which may open ways to attack the system. Based on the functionality of the registrar, would this information provisioning not something we would expect by a registrar and not be the pledge itself? This would also provide the option to provide this information to authenticated requestors only.

**Comment by mcr** [comment]
I think it would be worth repeating the information.  I'm unclear right now how the other telemetry is returned, so I need to look that part up.

**Comment by stfries** [comment]
Esko's comment to section 5.5.5:
1.	Change name “bootstrapping” in the request to more simpler “bootstrap” (also avoids the 1 or 2 p’s choice)
2.	Consider that we could split up the “error” status into two separate ones, e.g. “error-voucher” and “error-enroll” that indicate roughly at what phase the error occurred. This helps also the system/operator to interpret the reason and reason-context better, similar to how in BRSKI the status messages are processed – the processor knows whether the error occurred in the voucher-processing or in the enroll-processing so the “reason” and “reason-context” may not repeat that context information there.   So also helpful for the BRSKI PRM case if we can reuse the same messages/codes so it would need the greater error context to exist also outside the “reason” and “reason-context” fields in a standardized way. This is what “error-voucher” and “error-enroll” could provide.

**Comment by stfries** [comment]
Point 1 agreed, 
Point 2 may need some discussion. In BRSKI (and also in BRSKI-PRM) the specific status information is provided via two distinct endpoints. This may already ease the processing on the receiver side. In that aspect, it could be helpful to have also the status enumeration contain two distinct errors, which would avoid the parsing of the reason code. The general error should be kept additionally, to allow reuse of the mechanism for potential other errors, which are not defined in BRSKI-PRM. So there is at least the proposal to consider "error", "error-voucher", "error-enroll"

**Comment by siethower** [comment]
We reworked "5.5.5.  Request pledge-status".
@mcr and @EskoDijk please check and provide your feedback.

**Comment by EskoDijk** [comment]
I checked a bit and point 1 looks ok now. 
For point 2., a possibility / alternative is to include the complete 
* "voucher status" object into the JSON as a value , in case of stuck due ot a voucher-processing error, with a key name like "voucher-status".
* "enrollment status" object into the JSON as a value , in case of stuck due to a enrollment/LDevID-processing error, with a key name like "enroll-status".

Benefit of this is that potentially also both can be included if the Pledge e.g. detects an issue in the Voucher and also something's wrong in the LDevID.   A good Pledge may do a prelim check already of the LDevID even if the Voucher checks fail, to save time on the Commissioner person for a future attempt.

And it has the benefit that both already-defined objects can be completely re-used code-wise.  Disadvantage is a slightly (few bytes?) higher overhead.

**Comment by stfries** [comment]
The intention with the additional status update was to avoid storage of this information additionally on the pledge. In the positive case of the voucher-status and the enroll-status the status is straight forward by fetching the locally available information, without storing the previous status object. In the error case, we intended to use the error in the enumeration and provide the description what went wrong. That text may need to be updated.

**Comment by EskoDijk** [comment]
@stfries Sounds good - going back to the original issue description, it is worth to consider adding another pair of states. This comes from a practical use case point of view, where having a new device enrolled is nice but in practice it may still be unable to connect to a / its network for a variety of reasons.

So after the device successfully gets its LDevID, it needs to find the local IT network or room network (e.g. wireless) and connect to it. This connection process may fail somewhere. Now the installer might be sent to check on the new device with the handheld tool, in case the new device isn't reporting itself on the IT network.

Having some diagnosis info in this case would be helpful possibly to see why the device couldn't connect. (E.g. maybe the wireless coverage first needs to be improved, or a wireless AP is found but is for another domain , etc.)

Something like
* connect-success
* connect-error

In case the device didn't even attempt to connect yet (e.g. it was instructed to wait 2 weeks with that? Or another reason, e.g. no networks found at all) it stays in state "voucher-success" until it's time to connect.
Being in one of the "connect-*" states implies enroll-success has been transitioned before, just like enroll-* states imply that voucher-success has been transitioned before.

**Comment by stfries** [comment]
@EskoDijk Understand the approach. It may be a borderline case, as the pledge is expected to finish the enrollment also via the registrar-agent. That means if the enrollment went fine, the pledge would stay in enroll-success. That would provide the information to the service technician regarding the result of the bootstrapping (failure or success). Connectivity state information could be covered also by the application/process following bootstrapping. That's why I thought borderline. It is fine to add the two options as well, to be able to address connectivity problems after bootstrapping.

**Comment by EskoDijk** [comment]
@stfries It's true that these 2 states may not be used by a particular application ecosystem, e.g. if they define their own methods. Or they define an additional "status" resource that a portable tool can use to extract the information about the device post-enrollment.  So it's indeed more a concern for after enrollment. (Though getting on the local network is an important next step after enrollment, and can be seen as the final part of the bootstrapping process.)

But if we do provide these states it may be handy for an ecosystem adopting the solution, since they don't have to define a completely new method to check connectivity but can just plug into the existing framework. Due to the open nature of the JSON data, also ecosystem-specific data could always be added to the status information.

**Comment by stfries** [comment]
@EskoDijk Added in the last commit as discussed. Issue can be closed from my understanding.

**Comment by EskoDijk** [comment]
Ok, thanks & let's close it.
